File: client/src/sim/world.ts
import { Radians, Rect } from "./geometry";
import { Velocity } from "./physics";

export type ID = number;

export type GenNewID = () => ID;

export type Health = {
    max: number;
    current: number;
}

export enum BehaviourType {
    Actor = "Actor",
    Projectile = "Projectile",
}

export enum ModelType {
    Human = "Human",
    Monster = "Monster",
    Projectile = "Projectile",
}


export type Entity = {
    id: ID;
    health: Health;
    boundaries: Rect;
    rotation: Radians;
    model_type: ModelType;
    behaviour_type: BehaviourType,
    player_id: ID,
}

export type Player = {
    id: ID;
}

export type Process = {
    id: ID,
    entity_id: ID,
    payload: ProcessPayload,
}

export type ProcessPayload =
    | {
        type: "EntityMove";
        direction: Radians;
        velocity: Velocity;
    }
    | {
        type: "EntityShoot";
        cooldown: number;
        current_cooldown: number;
    }

export type WorldState = {
    boundaries: Rect;
    players: Map<ID, Player>;
    entities: Map<ID, Entity>;
    processes: Map<ID, Process>;
}

File: packages/game-5-sim/src/world.rs
use hashbrown::HashMap;
use crate::geometry::{Rect, Radians};
use crate::physics::{Velocity};

pub type ID = i32;

pub type GenNewID = Fn() -> ID;

#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
pub struct Health {
    pub max: u32,
    pub current: u32,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
pub enum BehaviourType {
    Actor,
    Projectile,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
pub enum ModelType {
    Human,
    Monster,
    Projectile,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
pub struct Entity {
    pub id: ID,
    pub health: Health,
    pub boundaries: Rect,
    pub rotation: Radians,
    pub model_type: ModelType,
    pub behaviour_type: BehaviourType,
    pub player_id: ID,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
pub struct Player {
    pub id: ID,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
pub struct Process {
    pub id: ID,
    pub entity_id: ID,
    pub payload: ProcessPayload,
}

#[derive(Debug, Copy, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum ProcessPayload {
    EntityMove {
        direction: Radians,
        velocity: Velocity,
    },
    EntityShoot {
        cooldown: i32,
        current_cooldown: i32,
    },
}

impl ProcessPayload {
    pub fn is_entity_move (&self) -> bool {
        match self {
            ProcessPayload::EntityMove { .. } => true,
            _ => false,
        }
    }

    pub fn is_entity_shoot (&self) -> bool {
        match self {
            ProcessPayload::EntityShoot { .. } => true,
            _ => false,
        }
    }
}

pub struct WorldState {
    pub new_id: ID,
    pub boundaries: Rect,
    pub players: HashMap<ID, Player>,
    pub entities: HashMap<ID, Entity>,
    pub processes: HashMap<ID, Process>,
}
